docs/cudf/source/developer_guide/testing.md

```diff
@@ -0,0 +1,105 @@
+# Testing cuDF
+
+## Tooling
+Tests in cuDF are written using [`pytest`](https://docs.pytest.org/en/latest/).
+Test coverage is measured using [`coverage.py`](https://coverage.readthedocs.io/en/6.4.1/),
+specifically the [`pytest-cov`](https://github.com/pytest-dev/pytest-cov) plugin.
+Code coverage reports are uploaded to [Codecov](https://app.codecov.io/gh/rapidsai/cudf).
+Each PR also indicates whether it increases or decreases in test coverage.
+
+## Test organization
+
+How tests are organized depends on which of the following two groups they fall into:
+
+1. Methods of classes like `DataFrame` or `Series`.
+2. Free functions operating on the above classes like `cudf.merge`.
+
+The former should be organized into directories named for the class (all lowercase), e.g. `dataframe/`.
+
+**Open question**: How should we group functionality within a class? How should we organize free functions? Some options:
+1. Match the grouping in our docs. It's somewhat arbitrary, but better than nothing.
+2. One file per function. This will create _a lot_ of files, but given how many tests we have that might not be crazy.
+3. Comparisons to pandas vs other. In some ways this grouping could be helpful, but I suspect it will only exacerbate our current problem of not knowing where to find tests. We need some sort of grouping by functionality.
+
+
+## Test contents
+
+### Writing tests 
+
+**Open question**: Do the guidelines below apply to pandas comparisons? For pandas comparisons it is IMO far less problematic to include standard and exceptional cases in the same tests.
+
+
+In general, functionality must be tested both for standard use cases and exceptional cases.
+Standard use cases may be covered using parametrization (using `pytest.mark.parametrize`).
+These use cases may also use the generic set of fixtures we provide (see **Add link to section**)
+Tests of standard use cases should typically include some coverage of:
+- Different dtypes, including nested dtypes (especially strings)
+- Mixed objects, e.g. binary operations between `DataFrame` and `Series`
+- Operations on scalars
+- Verifying all combinations of parameters for complex APIs like `cudf.merge`.
+
+Here are some of the most common exceptional cases to test:
```

Comment:
Some others I've encountered in pandas:

1. For numeric

- monotonically increasing
- monotonically decreasing
- all 0
- all 1
- random
- contains inf/-inf
- partially null
- integers at the implementation limit e.g. 2,147,483,647 for int32

2. Non en_US locale (affects strings the most)

Reply:
Thanks! That's super helpful.